File: gifLib/src/main/java/com/github/forjrking/gifLib/DrawableBytesTranscoder.java
package com.github.forjrking.gifLib;

import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.engine.Resource;
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapResource;
import com.bumptech.glide.load.resource.transcode.ResourceTranscoder;

import pl.droidsonroids.gif.GifDrawable;

public final class DrawableBytesTranscoder implements ResourceTranscoder<Drawable, byte[]> {
    private final BitmapPool bitmapPool;
    private final ResourceTranscoder<Bitmap, byte[]> bitmapBytesTranscoder;
    private final ResourceTranscoder<com.bumptech.glide.load.resource.gif.GifDrawable, byte[]> gifDrawableBytesTranscoder;
    // DES: 此处为自己的GifDrawable 插入后提供转换功能
    private final ResourceTranscoder<GifDrawable, byte[]> gifLibTranscoder;

    public DrawableBytesTranscoder(
            @NonNull BitmapPool bitmapPool,
            @NonNull ResourceTranscoder<Bitmap, byte[]> bitmapBytesTranscoder,
            @NonNull ResourceTranscoder<com.bumptech.glide.load.resource.gif.GifDrawable, byte[]> gifDrawableBytesTranscoder,
            @NonNull ResourceTranscoder<GifDrawable, byte[]> gifLibTranscoder) {
        this.bitmapPool = bitmapPool;
        this.bitmapBytesTranscoder = bitmapBytesTranscoder;
        this.gifDrawableBytesTranscoder = gifDrawableBytesTranscoder;
        this.gifLibTranscoder = gifLibTranscoder;
    }

    @Nullable
    @Override
    public Resource<byte[]> transcode(@NonNull Resource<Drawable> toTranscode,
                                      @NonNull Options options) {
        Drawable drawable = toTranscode.get();
        if (drawable instanceof BitmapDrawable) {
            return bitmapBytesTranscoder.transcode(
                    BitmapResource.obtain(((BitmapDrawable) drawable).getBitmap(), bitmapPool), options);
        } else if (drawable instanceof GifDrawable) {
            return gifLibTranscoder.transcode(toGifLibDrawableResource(toTranscode), options);
        } else if (drawable instanceof com.bumptech.glide.load.resource.gif.GifDrawable) {
            return gifDrawableBytesTranscoder.transcode(toGifDrawableResource(toTranscode), options);
        }
        return null;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    private static Resource<GifDrawable> toGifLibDrawableResource(@NonNull Resource<Drawable> resource) {
        return (Resource<GifDrawable>) (Resource<?>) resource;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    private static Resource<com.bumptech.glide.load.resource.gif.GifDrawable> toGifDrawableResource(@NonNull Resource<Drawable> resource) {
        return (Resource<com.bumptech.glide.load.resource.gif.GifDrawable>) (Resource<?>) resource;
    }
}

File: gifLib/src/main/java/com/github/forjrking/gifLib/DrawableBytesTranscoder.kt
package com.github.forjrking.gifLib

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.engine.Resource
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapResource
import com.bumptech.glide.load.resource.gif.GifDrawable
import com.bumptech.glide.load.resource.transcode.ResourceTranscoder

/**Drawable to Bytes*/
@Suppress("UNCHECKED_CAST")
class DrawableBytesTranscoder(
        private val bitmapPool: BitmapPool,
        private val bitmapBytesTranscoder: ResourceTranscoder<Bitmap, ByteArray>,
        private val gifDrawableBytesTranscoder: ResourceTranscoder<GifDrawable, ByteArray>,
        // DES: 此处为自己的GifDrawable 插入后提供转换功能
        private val gifLibTranscoder: ResourceTranscoder<pl.droidsonroids.gif.GifDrawable, ByteArray>) : ResourceTranscoder<Drawable, ByteArray> {

    override fun transcode(toTranscode: Resource<Drawable?>,
                           options: Options): Resource<ByteArray>? {
        return when (val drawable = toTranscode.get()) {
            is BitmapDrawable -> {
                bitmapBytesTranscoder.transcode(BitmapResource.obtain(drawable.bitmap, bitmapPool)!!, options)
            }
            is pl.droidsonroids.gif.GifDrawable -> {
                gifLibTranscoder.transcode(toGifLibDrawableResource(toTranscode), options)
            }
            is GifDrawable -> {
                gifDrawableBytesTranscoder.transcode(toGifDrawableResource(toTranscode), options)
            }
            else -> null
        }
    }

    companion object {
        private fun toGifLibDrawableResource(resource: Resource<Drawable?>): Resource<pl.droidsonroids.gif.GifDrawable> {
            return resource as Resource<pl.droidsonroids.gif.GifDrawable>
        }

        /** */
        private fun toGifDrawableResource(resource: Resource<Drawable?>): Resource<GifDrawable> {
            return resource as Resource<GifDrawable>
        }
    }

}